nifi-nar-bundles/nifi-elasticsearch-bundle/nifi-elasticsearch-processors/src/main/java/org/apache/nifi/processors/elasticsearch/PutElasticsearch.java

```diff
@@ -0,0 +1,231 @@
+        return Collections.unmodifiableList(descriptors);
+    }
+
+
+    @Override
+    public void onTrigger(final ProcessContext context, final ProcessSession session) throws ProcessException {
+        // Create the client if one does not already exist
+        createClient(context);
+        final int batchSize = context.getProperty(BATCH_SIZE).asInteger();
+        final String index = context.getProperty(INDEX).evaluateAttributeExpressions().getValue();
+        final String id_attribute = context.getProperty(ID_ATTRIBUTE).getValue();
+        final String docType = context.getProperty(TYPE).evaluateAttributeExpressions().getValue();
+
+        final List<FlowFile> flowFiles = session.get(batchSize);
+        if (flowFiles.isEmpty()) {
+            return;
+        }
+
+        final ProcessorLog logger = getLogger();
+
+        try {
+            final BulkRequestBuilder bulk = esClient.get().prepareBulk();
+            for (FlowFile file : flowFiles) {
+                final String id = file.getAttribute(id_attribute);
+                if (id == null) {
+                    getLogger().error("no value in identifier attribute {}", new Object[]{id_attribute});
+                    throw new ProcessException("No value in identifier attribute " + id_attribute);
+                }
+                session.read(file, new InputStreamCallback() {
+                    @Override
+                    public void process(final InputStream in) throws IOException {
+                        String json = IOUtils.toString(in, StandardCharsets.UTF_8)
+                                .replace("\r\n", " ").replace('\n', ' ').replace('\r', ' ');
+                        bulk.add(esClient.get().prepareIndex(index, docType, id)
+                                .setSource(json.getBytes(StandardCharsets.UTF_8)));
+                    }
+                });
+            }
+
+            final BulkResponse response = bulk.execute().actionGet();
+            if (response.hasFailures()) {
+                for (final BulkItemResponse item : response.getItems()) {
+                    final FlowFile flowFile = flowFiles.get(item.getItemId());
+                    if (item.isFailed()) {
+                        logger.error("Failed to insert {} into Elasticsearch due to {}",
+                                new Object[]{flowFile, item.getFailure()}, new Exception());
+                        session.transfer(flowFile, REL_FAILURE);
+
+                    } else {
+                        session.transfer(flowFile, REL_SUCCESS);
+
+                    }
+                }
+            } else {
+                for (final FlowFile flowFile : flowFiles) {
+                    session.transfer(flowFile, REL_SUCCESS);
+                }
+            }
+
+
+        } catch (NoNodeAvailableException nne) {
+            logger.error("Failed to insert {} into Elasticsearch No Node Available {}", new Object[]{nne}, nne);
+            for (final FlowFile flowFile : flowFiles) {
```

Comment:
For each of these cases, we can just call session.transfer(flowFile, REL_RETRY), rather than iterating over the FlowFiles.